File: BasicLib/SlidingTextWindow.cpp
#include "SlidingTextWindow.hpp"

#include <utility>
#include <cassert>

namespace funny
{
	SlidingTextWindow::SlidingTextWindow(StringType str)
		:lineContent_{ std::move(str) }
	{

	}

	void SlidingTextWindow::AdvanceChar(std::size_t n) noexcept
	{
		pos_ += n;
		assert(pos_ <= lineContent_.length() && pos_ >= 0);
	}

	bool SlidingTextWindow::AdvanceIfMatches(const StringType & str) noexcept
	{
		auto len = str.length();
		for (std::size_t i = 0;i < len;++i)
			if (PeekChar(i) != str[i]) return false;
		AdvanceChar(len);
		return true;
	}

	void SlidingTextWindow::Reset(std::size_t pos) noexcept
	{
		assert(pos <= lineContent_.length());
		pos_ = pos;
	}

	CharType SlidingTextWindow::PeekChar(std::size_t delta) noexcept
	{
		auto tmp = pos_;
		AdvanceChar(delta);
		CharType ret;
		if (pos_ >= lineContent_.length())
			ret = InvalidCharacter;
		else
			ret = lineContent_[pos_];
		Reset(tmp);
		return ret;
	}

	CharType SlidingTextWindow::NextChar() noexcept
	{
		auto c = PeekChar();
		if (c != InvalidCharacter)
			AdvanceChar();
		return c;
	}

	bool SlidingTextWindow::AdvanceIfPositiveInteger(int & result) noexcept
	{
		CharType c;
		std::size_t i{};
		int res{};
		c = PeekChar(i);
		while (CharUtility::IsDigit(c))
		{
			res = 10 * res + CharUtility::ToInt(c);
			c = PeekChar(++i);
		}
		AdvanceChar(i);
		result = res;
		return i != 0;
	}

	bool SlidingTextWindow::AdvanceIfMatches(CharType c) noexcept
	{
		if (PeekChar() == c)
		{
			AdvanceChar();
			return true;
		}
		return false;
	}

}

File: Compilers/RegexChart/RegexParser/SlidingTextWindow.cs
using System.Diagnostics;
using System.Text;

namespace RegexChart.RegexParser
{
    class SlidingTextWindow
    {
        public const char InvalidCharacter = char.MaxValue;

        public int Pos => _offset;

        private int _offset;
        private string _sourceText;

        public SlidingTextWindow(string source)
        {
            _sourceText = source;
        }

        public void AdvanceChar()
        {
            _offset++;
        }

        public void AdvanceChar(int n)
        {
            _offset += n;
            Debug.Assert(_offset >= 0 && _offset <= _sourceText.Length);
        }

        public char PeekChar()
        {
            if(_offset >= _sourceText.Length)
                return InvalidCharacter;          
            return _sourceText[_offset];
        }

        public void Reset(int position)
        {
            Debug.Assert(position <= _sourceText.Length, "Invalid position.");
            _offset = position;
        }

        public char PeekChar(int delta)
        {
            var tmp = _offset;
            this.AdvanceChar(delta);
            char ret;
            if (_offset >= _sourceText.Length)
                ret = InvalidCharacter;
            else
                ret = _sourceText[_offset];
            this.Reset(tmp);
            return ret;
        }

        public char NextChar()
        {
            var c = PeekChar();
            if (c != InvalidCharacter)
                this.AdvanceChar();
            return c;
        }

        public bool AdvanceIfMatches(char c)
        {
            if (PeekChar() == c)
            {
                AdvanceChar();
                return true;
            }
            return false;
        }

        public bool AdvanceIfMatches(string desired)
        {
            var length = desired.Length;
            for (int i = 0; i < length; i++)
            {
                if (PeekChar(i) != desired[i])
                    return false;
            }
            AdvanceChar(length);
            return true;
        }

        public bool AdvanceIfOneOf(string seq,out char ans)
        {
            var c = PeekChar();
            foreach(var c2 in seq)
            {
                if(c == c2)
                {
                    AdvanceChar();
                    ans = c;
                    return true;
                }
            }
            ans = default(char);
            return false;
        }

        public bool AdvanceIfPositiveInteger(out int result)
        {
            result = 0;
            int i = 0;
            char c;
            while(true)
            {
                c = PeekChar(i);
                if (char.IsDigit(c))
                    result = 10 * result + (int)char.GetNumericValue(c);               
                else break;
                i++;
            }
            AdvanceChar(i);
            return i != 0;   
        }

        public bool AdvanceIfName(out string result)
        {
            var sb = new StringBuilder();
            result = null;
            int i = 1;
            char c = PeekChar();
            //以下划线或字母开头
            if (char.IsLetter(c) || c == '_')
            {
                sb.Append(c);
                while (true)
                {
                    c = PeekChar(i);
                    if (char.IsLetterOrDigit(c) || c == '_')
                        sb.Append(c);
                    else break;
                    i++;
                }
            }
            else return false;
            result = sb.ToString();
            AdvanceChar(i);
            return true;
        }

        public bool IsValid(char c)
        {
            return c != InvalidCharacter;
        }

        public bool HasMoreChars()
        {
            return _offset < _sourceText.Length;
        }
    }


}
